main/src/main/scala/sbt/internal/server/BuildServerProtocol.scala

```diff
@@ -380,6 +431,76 @@ object BuildServerProtocol {
     }
   }
 
+  private val jsonParser: Parser[Try[JValue]] = (Parsers.any *)
+    .map(_.mkString)
+    .map(JsonParser.parseFromString)
+
+  private def bspRunTask: Def.Initialize[InputTask[Unit]] = Def.inputTaskDyn {
+    val runParams = jsonParser
+      .map(_.flatMap(json => Converter.fromJson[RunParams](json)))
+      .parsed
+      .get
+    val defaultClass = Keys.mainClass.value
+    val defaultJvmOptions = Keys.javaOptions.value
+
+    val mainClass = runParams.dataKind match {
+      case Some("scala-main-class") =>
+        val data = runParams.data.getOrElse(JNull)
+        Converter.fromJson[ScalaMainClass](data) match {
+          case Failure(e) =>
+            throw LangServerError(
+              ErrorCodes.ParseError,
+              e.getMessage
+            )
+          case Success(value) => value
+        }
+
+      case Some(dataKind) =>
+        throw LangServerError(
+          ErrorCodes.InvalidParams,
+          s"Unexpected data of kind '$dataKind', 'scala-main-class' is expected"
+        )
+
+      case None =>
+        ScalaMainClass(
+          defaultClass.getOrElse(
+            throw LangServerError(
+              ErrorCodes.InvalidParams,
+              "No default main class is defined"
+            )
+          ),
+          runParams.arguments,
+          defaultJvmOptions.toVector
+        )
+    }
+
+    runMainClassTask(mainClass, runParams.originId)
+  }
+
+  private def runMainClassTask(mainClass: ScalaMainClass, originId: Option[String]) = Def.task {
+    val state = Keys.state.value
+    val logger = Keys.streams.value.log
+    val classpath = Attributed.data(fullClasspath.value)
+    val forkOpts = ForkOptions(
+      javaHome = javaHome.value,
+      outputStrategy = outputStrategy.value,
+      // bootJars is empty by default because only jars on the user's classpath should be on the boot classpath
+      bootJars = Vector(),
+      workingDirectory = Some(baseDirectory.value),
+      runJVMOptions = mainClass.jvmOptions,
+      connectInput = connectInput.value,
+      envVars = envVars.value
+    )
+    val runner = new ForkRun(forkOpts)
+    val statusCode = runner
+      .run(mainClass.`class`, classpath, mainClass.arguments, logger)
+      .fold(
+        _ => StatusCode.Error,
+        _ => StatusCode.Success
+      )
```

Comment:
Would this block the entire server?
I created bgRun (or background run) for this type of use case - https://github.com/sbt/sbt/pull/2936

Reply:
This is tricky because I want to send a `statusCode` that depends on the return code of the execution. If I use `bgRun` then the task would terminate before the execution but I don't want to send the `Done` response to the client.

I can give it a try though.

Reply:
Also I cannot use `bgRun` implementation as it is because I need to overwrite the java options in the `ScalaRun` and the callback of the execution. IMO it is safer to have a separate implementation for `buildTarget/run`.

Reply:
ok. I'll accept this PR as-is, and maybe we can iterate on this as we learn more.

Reply:
Thank you! I have good hope that it will be improved.